src/main/java/org/jberet/support/io/DynamoDbItemReader.java

```diff
@@ -0,0 +1,267 @@
+	 */
+	@Inject
+	@BatchProperty
+	protected Integer end;
+	/**
+	 * Maximum number of items to read on server side
+	 */
+	@Inject
+	@BatchProperty
+	protected Integer limit;
+	/**
+	 * Filter expression.
+	 * Example: <code>price>=:minPrice and active=:active and #type=:type</code>
+	 */
+	@Inject
+	@BatchProperty
+	protected String filterExpression;
+	/**
+	 * Field nama aliases to inject in the {@link #filterExpression}.
+	 * This is a key-value map encoded in JSON.
+	 * Example: <code>{"#type":"type"}</code>
+	 */
+	@Inject
+	@BatchProperty
+	protected String filterExpressionNames;
+	/**
+	 * Values to inject in the {@link #filterExpression}.
+	 * This is a key-value map encoded in JSON.
+	 * Example: <code>{":minPrice":12,":active":true,":type":"candy"}</code>
+	 */
+	@Inject
+	@BatchProperty
+	protected String filterExpressionValues;
+	/**
+	 * Attributes to project in result bean.
+	 * This is a coma separated list of field names.
+	 * Example: <code>name,minPrice,active,type</code>
+	 */
+	@Inject
+	@BatchProperty
+	protected String attributesToProject;
+	/**
+	 * Current item iterator
+	 */
+	private Iterator<D> itemIterator;
+
+	@Override
+	public void open(Serializable serializable) throws Exception {
+		initEnhancedClient();
+		PageIterable<D> pageIterable = read(getTable());
+		Stream<D> itemStream = pageIterable.items().stream();
+		if (this.start != null && this.start >= 0) {
+			itemStream = itemStream.skip(this.start);
+		}
+		if (this.end != null && this.end > 0) {
+			itemStream = itemStream.limit(this.end - (this.start == null ? 0 : this.start));
+		}
+		itemIterator = itemStream.iterator();
+	}
+
+	private PageIterable<D> read(DynamoDbTable<D> table) {
+		Key queryKey = getQueryKey();
+		PageIterable<D> pageIterable;
+		if (this.indexName == null) {
+			if (queryKey == null) {
+				pageIterable = table.scan(this::configureScanRequest);
+			} else {
+				pageIterable = table.query(builder -> configureQueryRequest(builder, queryKey));
+			}
+		} else {
+			DynamoDbIndex<D> index = table.index(this.indexName);
+			if (queryKey == null) {
+				pageIterable = PageIterable.create(index.scan(this::configureScanRequest));
+			} else {
+				pageIterable = PageIterable.create(index.query(builder -> configureQueryRequest(builder, queryKey)));
+			}
+		}
+		return pageIterable;
+	}
+
+	private QueryEnhancedRequest.Builder configureQueryRequest(QueryEnhancedRequest.Builder builder, Key queryKey) {
+		return builder
+				.consistentRead(this.consistentRead)
+				.limit(this.limit)
+				.queryConditional(QueryConditional.keyEqualTo(queryKey))
+				.filterExpression(getFilterExpression());
+	}
+
+	private ScanEnhancedRequest.Builder configureScanRequest(ScanEnhancedRequest.Builder builder) {
+		return builder
+				.consistentRead(this.consistentRead)
+				.limit(this.limit)
+				.filterExpression(getFilterExpression())
+				.attributesToProject(getAttributesToProject());
+	}
+
+
+	@Override
+	public void close() throws Exception {
+
+	}
```

Comment:
the close method in both reader and writer classes are noop. Is there a need to close the underlying connection and resource to prevent resource leaking? I looked at the `EnhancedClient` but didn't see any close method.

Reply:
AFAIK there is nothing to do to free DynamoDB client resources because DynamoDBClient is a stateless HTTP based client.

That being said, in real world situation I advise to inject the DynamoDBClient with CDI and let CDI manage its lifecycle as well. This allows to share the same client for multiple readers/writers.

This PR is based on code running in production for a couple of months and based on [Quarkus Amazon Service](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-dynamodb.html)